plutus-ledger/src/Ledger/TimeSlot.hs

```diff
@@ -1,45 +1,111 @@
-{-# LANGUAGE NoImplicitPrelude #-}
+{-# LANGUAGE DeriveAnyClass     #-}
+{-# LANGUAGE DeriveGeneric      #-}
+{-# LANGUAGE DerivingStrategies #-}
+{-# LANGUAGE NamedFieldPuns     #-}
+{-# LANGUAGE NoImplicitPrelude  #-}
+{-# LANGUAGE TemplateHaskell    #-}
 
 -- This GHC option prevents the error:
 -- "GHC Core to PLC plugin: E042:Error: Unsupported feature: Kind: *"
 -- Because Plutus can't handle unboxed tuples which come from worker/wrapper
 {-# OPTIONS_GHC -fno-worker-wrapper #-}
+{-# OPTIONS_GHC -Wno-orphans        #-}
 
 module Ledger.TimeSlot(
-  slotRangeToPOSIXTimeRange
-, slotToPOSIXTime
+  SlotConfig(..)
+, slotRangeToPOSIXTimeRange
+, slotToPOSIXTimeRange
+, slotToBeginPOSIXTime
+, slotToEndPOSIXTime
 , posixTimeRangeToSlotRange
 , posixTimeToSlot
+, currentSlot
 ) where
 
-import           Plutus.V1.Ledger.Slot (Slot (Slot), SlotRange)
-import           Plutus.V1.Ledger.Time (POSIXTime (POSIXTime), POSIXTimeRange)
-import           PlutusTx.Prelude
+import           Data.Aeson                (FromJSON, ToJSON)
+import           Data.Default              (Default (def))
+import qualified Data.Time.Clock           as Time
+import qualified Data.Time.Clock.POSIX     as Time
+import           GHC.Generics              (Generic)
+import           Plutus.V1.Ledger.Interval (Interval (Interval, ivFrom, ivTo), interval)
+import           Plutus.V1.Ledger.Slot     (Slot (Slot), SlotRange)
+import           Plutus.V1.Ledger.Time     (POSIXTime (POSIXTime, getPOSIXTime), POSIXTimeRange)
+import           PlutusTx.Lift             (makeLift)
+import           PlutusTx.Prelude          hiding (Eq, (<$>))
+import           Prelude                   (Eq, IO, Show, (<$>))
+import qualified Prelude                   as Haskell
 
-{-# INLINABLE shelleyLaunchDate #-}
--- | 'shelleyLaunchDatePOSIXTime' corresponds to the time 2020-07-29T21:44:51Z
--- which is 1596059091 in POSIX time (number of seconds since
--- 1970-01-01T00:00:00Z).
-shelleyLaunchDate :: Integer
-shelleyLaunchDate = 1596059091
+-- | Datatype to configure the length of one slot and the beginning of the
+-- first slot.
+data SlotConfig =
+    SlotConfig
+        { scSlotLength   :: Integer -- ^ Length (number of seconds) of 1 slot
```

Comment:
Would have used the `Second` datatype, but that does not work on the on chain part of Plutus. Normally, functions in this file shouldn't be used onchain, but Marlowe still does (added a TODO). When Marlowe updates their part of code, this file can be rewritten.